### ISSUE TITLE
[Question] How to customize decoding process (mask out unwanted symbol)?

### ISSUE THREAD

**Issue by mrzjy** [tensorflow/tensor2tensor#554]
So basically I trained my transformer model and it's able to generate some really good results. However, I don't want the "unknown" symbol (which is in the vocabulary) to appear in the generated results. Instead of removing the 'unk' from the vocabulary and retrain a new model, I intend to simply customize the decoding process : mask out its probability and make the model choose the second most probable word, which is a really simple idea.  

I would like to know if there is a way to achieve this, any help or suggestion is appreciated!

**Comment by 1nsunym** [comment]
Decoding methods are implemented in `utils/decoding.py`
You can go and explicitly modify `decoding.py` in `~/usr_env/lib/pythonX.X/site-packages/tensor2tensor/utils`(if you are using virtualenv) or wherever your tensor2tensor package is installed.

**Comment by mrzjy** [comment]
Thanks for your comment @1nsunym . However I think that directly deleting unknown symbols after decoding might harm the model's performance, in fact what I'd like to do is to get the softmax logits at every decoding time step and thus mask out certain word by setting its value to -∞. 

By checking the [utils/decoding.py](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/decoding.py), I think the solution should lie in this [line](https://github.com/tensorflow/tensor2tensor/blob/103d0571ef48d82b61aae67cd61d716738f9fcfe/tensor2tensor/utils/decoding.py#L260), where we define the prediction/decoding function, but I've no idea how to proceed.

**Comment by colmantse** [comment]
Do you mean censoring certain outputs during decoding? You should probably look at beam-search or greedy infer.

**Comment by 1nsunym** [comment]
Taking at look at [this function](  https://github.com/tensorflow/tensor2tensor/blob/103d0571ef48d82b61aae67cd61d716738f9fcfe/tensor2tensor/utils/beam_search.py#L408) might help.

**Comment by mrzjy** [comment]
@1nsunym  Thanks for the suggestion, I'll try to do this. By the way, how can I tell Tensor2tensor to use my customized beam_decode? Is there a registry way for that?

**Comment by mrzjy** [comment]
@colmantse Yes, I think you're right, thanks! I'll try to look at the beam search part.

**Comment by 1nsunym** [comment]
I'm only familiar with the hacky way sorry.

**Comment by mrzjy** [comment]
@1nsunym I finally succeed in modifying the code as you said, thanks a lot!
